client/my-sites/plans-features-main/components/plan-upsell-modal/index.tsx

```diff
@@ -0,0 +1,131 @@
+import { FreePlanPaidDomainDialog } from './free-plan-paid-domain-dialog';
+import PaidPlanIsRequiredDialog from './paid-plan-is-required-dialog';
+import type { DomainSuggestion } from '@automattic/data-stores';
+import type { DataResponse } from 'calypso/my-sites/plans-grid/types';
+
+export const PAID_PLAN_IS_REQUIRED_DIALOG = 'PAID_PLAN_IS_REQUIRED_DIALOG';
+export const FREE_PLAN_PAID_DOMAIN_DIALOG = 'FREE_PLAN_PAID_DOMAIN_DIALOG';
+export const FREE_PLAN_FREE_DOMAIN_DIALOG = 'FREE_PLAN_FREE_DOMAIN_DIALOG';
+export const MODAL_LOADER = 'MODAL_LOADER';
+
+export type ModalType =
+	| typeof FREE_PLAN_FREE_DOMAIN_DIALOG
+	| typeof FREE_PLAN_PAID_DOMAIN_DIALOG
+	| typeof PAID_PLAN_IS_REQUIRED_DIALOG
+	| typeof MODAL_LOADER;
+
+export type DomainPlanDialogProps = {
+	wpcomFreeDomainSuggestion: DataResponse< DomainSuggestion >;
+	suggestedPlanSlug: PlanSlug;
+	onFreePlanSelected: ( isDomainRetained?: boolean ) => void;
+	onPlanSelected: () => void;
+};
+
+type ModalContainerProps = {
+	isModalOpen: boolean;
+	paidDomainName?: string;
+	modalType?: ModalType | null;
+	wpcomFreeDomainSuggestion: DataResponse< DomainSuggestions.DomainSuggestion >;
+	onClose: () => void;
+	onFreePlanSelected: ( isDomainRetained?: boolean ) => void;
+	onPlanSelected: ( planSlug: string ) => void;
+};
+
+// See p2-pbxNRc-2Ri#comment-4703 for more context
+export const MODAL_VIEW_EVENT_NAME = 'calypso_plan_upsell_modal_view';
+
+function DisplayedModal( {
+	paidDomainName,
+	modalType,
+	wpcomFreeDomainSuggestion,
+	onFreePlanSelected,
+	onPlanSelected,
+}: Omit< ModalContainerProps, 'selectedPlan' > ) {
+	switch ( modalType ) {
+		case FREE_PLAN_PAID_DOMAIN_DIALOG:
+			return (
+				<FreePlanPaidDomainDialog
+					paidDomainName={ paidDomainName as string }
+					wpcomFreeDomainSuggestion={ wpcomFreeDomainSuggestion }
+					suggestedPlanSlug={ PLAN_PERSONAL }
+					onFreePlanSelected={ onFreePlanSelected }
+					onPlanSelected={ () => {
+						onPlanSelected( PLAN_PERSONAL );
+					} }
+				/>
+			);
+		case FREE_PLAN_FREE_DOMAIN_DIALOG:
+			return (
+				<FreePlanFreeDomainDialog
+					suggestedPlanSlug={ PLAN_PERSONAL }
+					wpcomFreeDomainSuggestion={ wpcomFreeDomainSuggestion }
+					onFreePlanSelected={ onFreePlanSelected }
+					onPlanSelected={ () => {
+						onPlanSelected( PLAN_PERSONAL );
+					} }
+				/>
+			);
+		case PAID_PLAN_IS_REQUIRED_DIALOG:
+			return (
+				<PaidPlanIsRequiredDialog
+					paidDomainName={ paidDomainName as string }
+					suggestedPlanSlug={ PLAN_PERSONAL }
+					wpcomFreeDomainSuggestion={ wpcomFreeDomainSuggestion }
+					onFreePlanSelected={ onFreePlanSelected }
+					onPlanSelected={ () => {
+						onPlanSelected( PLAN_PERSONAL );
+					} }
+				/>
+			);
+		default:
+			break;
+	}
+
+	return null;
+}
+
+export default function ModalContainer( props: ModalContainerProps ) {
+	const { isModalOpen, modalType } = props;
+	if ( ! isModalOpen || ! modalType ) {
+		return;
+	}
```

Comment:
I've used the modal before, but don't recall exactly how it worked underneath. Curious:

1. Instead of unmounting here, which would keep the `ModalContainer` in the tree with all its properties (not sure if that means something), why not unmount at the consuming end (plans-features-main)? `{ isModalOpen && <PlanUpsellModal ...`

2. Although, why not set the Dialog's `isVisible` property with this value instead? So we don't unmount anything on our end. Would that have the same effect, potentially a better/smoother experience? Or is that something that you want to do due to use of `Global` below - to make sure it's unmounted on close (and would Dialog/react-modal not unmount the children when "closed")?

Reply:
> Instead of unmounting here, which would keep the ModalContainer in the tree with all its properties (not sure if that means something), why not unmount at the consuming end (plans-features-main)

Initially I was trying to encapsulate the model visible state inside one this component but it became too complex and convoluted. This is a remnant of that approach.

>  but don't recall exactly how it worked underneath

React-modal uses portals so a simple css hiding broke the code in my first attempt and wouldn't work. As you can see below, when the component is unmounted the component is manually deleted from the DOM.

https://github.com/reactjs/react-modal/blob/master/src/components/Modal.js#L186
 

Thanks for recommending the isVisible approach it makes sense. There isn't a general smoothness of experience from my testing but I think that is the more idiomatic approach here given the library that we use under the hood.

Reply:
Actually upon trying this out right now, I remembered that I did attempt this approach earlier, But for some reason the unmounting doesn't happen. 
<img width="1057" alt="image" src="https://github.com/Automattic/wp-calypso/assets/3422709/6b665946-4ac3-4bf8-adc7-aad5ccc220b5">

For now I am just going to leave the existing approach since it works well. We can probably pick up the improvement as a follow up and a deeper investigation.

Reply:
@jdc91 Was the screenshot captured using the `isVisible` prop? I tried something like
```
diff --git a/client/my-sites/plans-features-main/components/plan-upsell-modal/index.tsx b/client/my-sites/plans-features-main/components/plan-upsell-modal/index.tsx
index 1dc174e250..63b877f2a7 100644
--- a/client/my-sites/plans-features-main/components/plan-upsell-modal/index.tsx
+++ b/client/my-sites/plans-features-main/components/plan-upsell-modal/index.tsx
@@ -91,9 +91,7 @@ function DisplayedModal( {
 
 export default function ModalContainer( props: ModalContainerProps ) {
 	const { isModalOpen, modalType } = props;
-	if ( ! isModalOpen || ! modalType ) {
-		return;
-	}
+	const isVisible = isModalOpen && !! modalType;
 
 	const modalWidth = () => {
 		switch ( modalType ) {
@@ -108,7 +106,7 @@ export default function ModalContainer( props: ModalContainerProps ) {
 	return (
 		<Dialog
 			isBackdropVisible={ true }
-			isVisible={ true }
+			isVisible={ isVisible }
 			onClose={ props.onClose }
 			showCloseIcon={ true }
 			labelledby="plan-upsell-modal-title"
```
and it works for me. I think I must have misunderstood what you meant by the things you were trying :)

Reply:
I might try and see what happens too, dig a little. 

p.s. this is also what's recommended in the docs too:

> The onClose property must be set, and should modify the parent's state such that the dialog's isVisible property will be false when render is called.
> 
> By controlling the dialog's visibility through the isVisible property, the dialog component itself is responsible for providing any CSS transitions to animate the opening/closing of the dialog. This also keeps the parent's code clean and readable, with a minimal amount of boilerplate code required to show a dialog.

So worth giving a try.